docs/design/proposals/references/README.md

```diff
@@ -0,0 +1,243 @@
+# [Design] Support Resource Relationships
+
+## Problem Description
+
+In micro service world, often users need to be using multiple services to run an application or Job. For example, to create a model in SageMaker, you could use a combination of SageMaker, IAM, S3, ECR and VPC services. Some of these resources need to be created in serial manner because output of one API(e.g. IAM role ARN, vpc-id) is an input to the another API. 
+
+As an ACK user, I want to enable to be able to express the relationship between resources in ACK CRs, so that I can submit multiple Specs related my application/infrastructure at once without waiting for another resources to get created.
+
+## Proposed Implementation
+Kubernetes style of expressing resource relationship is by using [object references](https://godoc.org/k8s.io/api/core/v1#ObjectReference), [typedLocalObjectReference](https://godoc.org/k8s.io/api/core/v1#TypedLocalObjectReference) or [localObjectReference](https://godoc.org/k8s.io/api/core/v1#LocalObjectReference). Some examples of this are [Secrets](https://github.com/kubernetes/api/blob/4e9f5db10201ce6270b160e96e3f99ddb580135e/core/v1/types.go#L861) and [configmaps](https://github.com/kubernetes/api/blob/4e9f5db10201ce6270b160e96e3f99ddb580135e/core/v1/types.go#L1999-L2016) in K8s API. In the following sections, we present how operator pattern can be used in conjunction with references in ACK to support relationsip between resources.
+
+### Modification to CRDs
+Currently, each top level resource in API model of the AWS service has a CRD. The field types in the CRD are structs, lists or primitive datatypes. A resource in K8s can be identified by a combination of metadata - `APIVersion, Kind, Name and Namespace`. Since AWS APIs have a deterministic input shape, APIVersion and Kind can be fixed at code generation time while Namespace and Name of resource will be a user input at runtime. To facilitate this, we propose to extend the K8s style of referencing resources in ACK by introducing a new datatype `AWSResourceReference` in `apis/core/v1alpha1/common.go`:
+
+```
+// AWSResourceReference represents a reference to another CR which refers to a AWS resource in Kubernetes 
+type AWSResourceReference struct {
+    // Namespace of the resource being referred to
+    Namespace string `json:"namespace"`
```

Comment:
In Kubernetes, namespaces are security boundary. So, the ability to reference an object in another namespace might result in unexpected consequences. In fact, this is one of the main reasons that managed resources are cluster-scoped in Crossplane, so that they can be shared and everyone knows they can be shared, similar to how you can use other team's resources in AWS console.

We actually had a [chat](https://kubernetes.slack.com/archives/C0EG7JC6T/p1611967173070000) in sig-api-machinery about this because I was looking for a more explicit statement in the API conventions and now there is an [open PR](https://github.com/kubernetes/community/pull/5455) to Kubernetes API conventions doc listing the reasons why references across namespaces should be avoided. A concrete example is the ownership relation built by object reference, i.e. `ownerReference`. It's [explicitly forbidden](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/) for that reference to span across namespaces.

It has some consequences in controller deployment as well. It allows a controller to see and fetch values from other resources that it otherwise might not have permission for. For example, I set up a SageMaker controller in my namespace `A` and reference a VPC in namespace `B`:
* The SageMaker controller has to have `VPC` RBAC permissions either cluster-wide or in namespace `B` so that it can query that custom resource.
* The IAM permissions of SageMaker controller may or may not cover the permissions to know about the `VPC` in namespace `B`, i.e. it could be in another AWS account or region or just simply no `describe-vpc` permission on SageMaker IAM role.

That kinds of impersonation on controller level has to happen for any kind of referencing mechanism since it's the controller that fetches the value, but in native Kubernetes controllers it's always either in the same namespace or namespace->cluster in order not to breach the namespace boundary expectation of users.

To conclude, I'd suggest having only the `Name` in this struct so that teams do not end up referencing resources in other namespaces while the owner of those namespaces feel a false sense of security.